File: modal/src/main/java/ir/drax/modal/model/ModalObj.java
package ir.drax.modal.model;

import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import androidx.lifecycle.LiveData;

import java.util.List;

public class ModalObj extends LiveData<ModalObj> {
    private CharSequence title="";
    private MoButton message=new MoButton("",0,null);
    private int icon;
    private MoButton reAction;
    private int direction = Direction.BottomToTop;
    private ViewGroup root;
    public static String ViewIdTag= ModalObj.class.getSimpleName();
    private boolean blurEnabled = false;
    private boolean lockVisibility=false;
    private Listener listener;
    private Enum<Modal.Type> type=Modal.Type.Alert;
    private View modal;
    private int contentView;
    private List<MoButton> list;
    private int progress = 0;
    private int changedIndex=-1;


    public void setTitle(CharSequence title) {
        this.title = title;
        postValue(setChangedIndex(0));
    }

    public void setMessage(MoButton message) {
        this.message = message;
        postValue(setChangedIndex(1));
    }

    public void setIcon(int icon) {
        this.icon = icon;
        postValue(setChangedIndex(2));
    }

    public void setReAction(MoButton reAction) {
        this.reAction = reAction;
        postValue(setChangedIndex(3));
    }

    public void setDirection(int direction) {
        this.direction = direction;
    }

    public void setRoot(ViewGroup root) {
        this.root = root;
    }

    public void setBlurEnabled(boolean blurEnabled) {
        this.blurEnabled = blurEnabled;
    }

    public void setLockVisibility(boolean lockVisibility) {
        this.lockVisibility = lockVisibility;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public void setType(Enum<Modal.Type> type) {
        this.type = type;
    }

    public void setModal(View modal) {
        this.modal = modal;
    }

    public CharSequence getTitle() {
        return title;
    }

    public MoButton getMessage() {
        return message;
    }

    public int getIcon() {
        return icon;
    }

    public MoButton getReAction() {
        return reAction;
    }

    public int getDirection() {
        return direction;
    }

    public ViewGroup getRoot() {
        return root;
    }

    public String getViewIdTag() {
        return ViewIdTag;
    }

    public boolean isBlurEnabled() {
        return blurEnabled;
    }

    public boolean isLockVisibility() {
        return lockVisibility;
    }

    public Listener getListener() {
        return listener;
    }

    public Enum<Modal.Type> getType() {
        return type;
    }

    public View getModal() {
        return modal;
    }

    public List<MoButton> getList() {
        return list;
    }

    public void setList(List<MoButton> list) {
        this.list = list;
        postValue(setChangedIndex(4));

    }

    public int getProgress() {
        return progress;
    }

    public void setProgress(int progress) {
        this.progress = progress;
        postValue(setChangedIndex(5));

    }

    private ModalObj setChangedIndex(int i){
        changedIndex=i;
        return this;
    }

    public int getChangedIndex() {
        return changedIndex;
    }

    @Override
    protected void onActive() {
        super.onActive();
        Log.e("onActive",""+hasActiveObservers());
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        Log.e("onInactive",""+hasActiveObservers());
    }

    public int getContentView() {
        return contentView;
    }

    public void setContentView(int contentView) {
        this.contentView = contentView;
    }
}

File: modal/src/main/java/ir/drax/modal/model/ModalObj.kt
package ir.drax.modal.model

import android.graphics.drawable.Drawable
import android.view.View
import androidx.lifecycle.MutableLiveData
import ir.drax.modal.Listener
import ir.drax.modal.Modal

data class ModalObj(
        private val source:ModalObj? = null,
        var type:Modal.Type = source?.type ?: Modal.Type.Alert
) {

    var listener:Listener?= source?.listener
    var modal:View?= source?.modal
    var contentView:View? = source?.contentView
    var onShow:()->Unit= source?.onShow ?: {}
    var onDismiss:()->Unit= source?.onDismiss ?:{}
    var blurEnabled:Boolean= source?.blurEnabled ?: false
    var direction:Modal.Direction = source?.direction ?: Modal.Direction.Bottom
    var lockVisibility:Boolean= source?.lockVisibility ?: false
    var title:CharSequence= source?.title ?: ""
    var message:MoButton = source?.message ?: MoButton("")
    var icon:Int = source?.icon?: 0
    var callback: MoButton? = source?.callback
    var list:List<MoButton> = source?.list ?: listOf()
    var progress:Int = source?.progress ?: 0
    var backgroundDrawable:Drawable? = source?.backgroundDrawable

    var animationDuration:Long = source?.animationDuration ?: 250L
    var animationStartDelay:Long = source?.animationStartDelay ?: 0L

    init {
        if (listener == null) {
            listener = object:Listener(){
                override fun onDismiss() {
                    super.onDismiss()
                    onDismiss.invoke()
                }

                override fun onShow() {
                    super.onShow()
                    onShow.invoke()
                }
            }
        }


    }


    var update = MutableLiveData<ModalObj>()

    class Builder{
        private val modalObj = ModalObj()

        fun blurEnabled(isBlurEffectEnabled: Boolean) = apply { this.modalObj.blurEnabled = isBlurEffectEnabled }
        fun listener(listener: Listener) = apply { this.modalObj.listener = listener }
        fun onShow(onShow: ()->Unit) = apply { this.modalObj.onShow = onShow }
        fun onDismiss(onDismiss: ()->Unit) = apply { this.modalObj.onDismiss = onDismiss }
        fun direction(direction: Modal.Direction) = apply { this.modalObj.direction = direction }
        fun lockVisibility(lockVisibility: Boolean) = apply { this.modalObj.lockVisibility = lockVisibility }
        fun title(title: CharSequence) = apply { this.modalObj.title = title }
        fun message(message: String) = apply { this.modalObj.message = MoButton(message) }
        fun message(message: MoButton) = apply { this.modalObj.message = message }
        fun icon(icon: Int) = apply { this.modalObj.icon = icon }
        fun callback(callback: MoButton) = apply { this.modalObj.callback = callback }

        fun build() = modalObj
    }

    companion object{
        @JvmField
        var VIEW_TAG_ID: String? = this.javaClass.canonicalName
    }


    fun copy() = ModalObj(source = this)

}